brave/browser/brave_browser_context.cc

```diff
@@ -363,12 +474,75 @@ void BraveBrowserContext::UpdateDefaultZoomLevel() {
       ->OnDefaultZoomLevelChanged();
 }
 
+void BraveBrowserContext::TorSetProxy(
+    scoped_refptr<brightray::URLRequestContextGetter>
+      url_request_context_getter,
+    const base::FilePath partition_path) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  if (!url_request_context_getter || !isolated_storage_)
+    return;
+  if (tor_proxy_.is_valid()) {
+    auto proxy_service = url_request_context_getter->GetURLRequestContext()->
+      proxy_service();
+    // Notice CreateRequestContextForStoragePartition will only be called once
+    // per partition_path so there is no need to cache password per origin
+    std::string origin = partition_path.DirName().BaseName().value();
+    std::unique_ptr<net::ProxyConfigServiceTor>
+      config(new net::ProxyConfigServiceTor(tor_proxy_));
+    config->SetUsername(origin);
+    proxy_service->ResetConfigService(std::move(config));
+  }
+}
+
 content::PermissionManager* BraveBrowserContext::GetPermissionManager() {
   if (!permission_manager_.get())
     permission_manager_.reset(new BravePermissionManager);
   return permission_manager_.get();
 }
 
+void BraveBrowserContext::LaunchTorProcess(base::WaitableEvent* tor_launched,
+                                           const std::string& tor_path,
+                                           const std::string& tor_host,
+                                           const std::string& tor_port) {
+  content::ServiceManagerConnection::GetForProcess()
+    ->GetConnector()
+    ->BindInterface(tor::mojom::kTorServiceName,
+                    &tor_launcher_);
+
+  tor_launcher_.set_connection_error_handler(
+    base::BindOnce(&BraveBrowserContext::OnTorLauncherCrashed,
+                   base::Unretained(this)));
+  std::vector<std::string> args;
+  args.push_back("--ignore-missing-torrc");
+  args.push_back("-f");
+  args.push_back("/nonexistent");
+  args.push_back("--defaults-torrc");
+  args.push_back("/nonexistent");
+  args.push_back("--SocksPort");
+  args.push_back(tor_host + ":" + tor_port);
+  base::FilePath user_data_dir;
+  PathService::Get(chrome::DIR_USER_DATA, &user_data_dir);
+  if (!user_data_dir.empty()) {
+    args.push_back("--DataDirectory");
+    args.push_back(user_data_dir.value());
```

Comment:
I think this should be `user_data_dir + "/tor/data"`, unless I misunderstand?  The tor daemon will assume it owns this directory, and may use it to store various files -- conceivably conflicting with Brave's ~/.config/brave hierarchy.